{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing GFF output from CodingQuarry\n",
    "\n",
    "**Author:** Darcy Jones <br />\n",
    "**Contact:** darcy.ab.jones@gmail.com <br />\n",
    "**Date:** 02/04/2015 <br />\n",
    "\n",
    "\n",
    "[CodingQuarry](https://sourceforge.net/projects/codingquarry/) is a gene prediction program optimised for annotating fungal genomes.\n",
    "We were having a problem extracting ORF sequences from the GFF output of CodingQuarry because it appears to code some exons as 'CDS' types.\n",
    "This notebook converts the `type` of multiple 'CDS' features with the same `parent` to 'exon' and recodes the `ID` appropriately.\n",
    "Then coding sequences of genes for each genome is extracted and saved as fasta files as nucleotide and amino acid sequences.\n",
    "\n",
    "## Required software\n",
    "\n",
    "This was run with Python 3.4.1 with the following packages:\n",
    "- bcbio-gff (0.6.2)\n",
    "- biopython (1.65)\n",
    "- certifi (14.5.14)\n",
    "- ipython (3.0.0)\n",
    "- Jinja2 (2.7.3)\n",
    "- jsonschema (2.4.0)\n",
    "- MarkupSafe (0.23)\n",
    "- pyzmq (14.5.0)\n",
    "- six (1.9.0)\n",
    "- tornado (4.1)\n",
    "- mistune (0.5.1)\n",
    "- Pygments (2.0.2)\n",
    "\n",
    "\n",
    "\n",
    "## Relevant literature\n",
    "\n",
    "Testa, A. C., Hane, J. K., Ellwood, S. R., and Oliver, R. P. 2015. [CodingQuarry: highly accurate hidden Markov model gene prediction in fungal genomes using RNA-seq transcripts.](http://www.biomedcentral.com/1471-2164/16/170) BMC Genomics. 16:1â€“12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.SeqFeature import CompoundLocation \n",
    "from BCBio import GFF\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def gff_correct(in_path, out_path):\n",
    "    \"\"\" Recode multiple CDS records as exons in GFF.\n",
    "\n",
    "    Keyword arguments:\n",
    "    in_path -- Path to input gff file (str, required)\n",
    "    out_path -- Path to write new GFF to (str, required)\n",
    "    \"\"\"\n",
    "    with open(in_gff, 'rU') as in_handle, \\\n",
    "            open(out_path, 'w') as out_handle:\n",
    "        records = GFF.parse(in_handle)\n",
    "\n",
    "        def gff_iterator(records):\n",
    "            \"\"\" Using a generator to speed up the process. \"\"\"\n",
    "            for record in records:\n",
    "                for feature in record.features:\n",
    "                    sub_features = feature._sub_features\n",
    "                    new_sub_features = list()\n",
    "                    sub_features_cds = [f for f in feature._sub_features\n",
    "                                        if f.type == 'CDS']\n",
    "                    new_sub_features.extend([f for f in feature._sub_features\n",
    "                                             if f.type != 'CDS'])\n",
    "                    if len(sub_features_cds) == 1:\n",
    "                        new_sub_features.extend(sub_features_cds)\n",
    "                    elif len(sub_features_cds) > 1:\n",
    "                        for sub_feature in sub_features_cds:\n",
    "                            sub_feature.type = 'exon'\n",
    "                            sub_feature.id = ('exon:' +\n",
    "                                              sub_feature.id.strip('CDS:'))\n",
    "                            sub_feature.qualifiers['ID'] = sub_feature.id\n",
    "                            del sub_feature.qualifiers['phase']\n",
    "                            new_sub_features.append(sub_feature)\n",
    "                    feature._sub_features = new_sub_features\n",
    "                yield record\n",
    "\n",
    "        output = GFF.write(gff_iterator(records), out_handle)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def subfeatures(feature):\n",
    "    \"\"\" Return a location object from GFF output.\n",
    "\n",
    "    The BCBio GFF parser adds exons and CDS's to features as sub_features\n",
    "    which has since been depreciated in Biopython in favour of the\n",
    "    CompoundLocation object. This function returns a Location object that\n",
    "    we can use to extract sequences.\n",
    "\n",
    "    Keyword arguments:\n",
    "    feature -- A SeqFeature record with the _sub_features attribute.\n",
    "\n",
    "    Returns:\n",
    "    An ExactLocation or CompundLocation object.\n",
    "    \"\"\"\n",
    "    sub_features_cds = [f for f in feature._sub_features\n",
    "                        if f.type == 'CDS']\n",
    "    sub_features_exons = [f for f in feature._sub_features\n",
    "                          if f.type == 'exon']\n",
    "    strand = feature.strand\n",
    "    if len(sub_features_exons) > 0:\n",
    "        locations = [f.location for f in sub_features_exons]\n",
    "        locations.sort(key=lambda l: l.start)\n",
    "        if strand == -1:\n",
    "            \"\"\" When calling CompoundLocation.extract() the sequences are\n",
    "            extracted in the order that they are encountered. For features on\n",
    "            the - strand, we need to reverse this order. \"\"\"\n",
    "            locations.reverse()\n",
    "        return CompoundLocation(locations)\n",
    "    elif len(sub_features_cds) > 0:\n",
    "        if len(sub_features_cds) > 1:\n",
    "            locations = [f.location for f in sub_features_cds]\n",
    "            locations.sort(key=lambda l: l.start)\n",
    "            if strand == -1:\n",
    "                locations.reverse()\n",
    "            locations = CompoundLocation(locations)\n",
    "        else:\n",
    "            locations = sub_features_cds[0]  # One CDS returns an ExactLocation\n",
    "        return locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def gff_to_genes(in_gff, in_genome):\n",
    "    \"\"\" Clean GFF files and extract sequences.\n",
    "\n",
    "    Keyword arguments:\n",
    "    in_gff -- Path to input GFF3 (str, required)\n",
    "    in_genome -- Path to fasta genome corresponding to `in_gff` (str, required)\n",
    "\n",
    "    Returns:\n",
    "    A tuple with 0 = A list of all extracted SeqRecords, and 1 = Extracted\n",
    "        SeqRecords with unexpected stop codons within the sequence.\n",
    "    \"\"\"\n",
    "    split_gff_path = os.path.splitext(in_gff)[0]  # Strip the extension\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(in_genome, format='fasta'))\n",
    "\n",
    "    with open(in_gff, 'rU') as in_handle:\n",
    "        genome_with_features = GFF.parse(in_handle, base_dict=genome)\n",
    "\n",
    "        sequences = list()\n",
    "        dubious_sequences = list()\n",
    "\n",
    "        for scaffold in genome_with_features:\n",
    "            for feature in scaffold.features:\n",
    "                subfeat = subfeatures(feature)\n",
    "                seq = subfeat.extract(genome[scaffold.id])\n",
    "                seq.id = feature.id\n",
    "                seq.name = feature.id\n",
    "                sequences.append(seq)\n",
    "\n",
    "                \"\"\" Test if there are unexpected stop codons. \"\"\"\n",
    "                if '*' in seq.seq.translate()[:-1]:\n",
    "                    dubious_sequences.append(seq)\n",
    "    return sequences, dubious_sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have the functions ready to process CodingQuarry's weird GFF's and extract protein coding sequences.\n",
    "Now we can loop through GFF's and their relevant genomes to perform those operations quickly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cds_suffix = '.genes.fna'\n",
    "orf_suffix = '.genes.faa'\n",
    "dubious_suffix = '.genes.weird_cds.fna'\n",
    "cleaned_suffix = '.cleaned'\n",
    "\n",
    "in_files = [\n",
    "    (\"9/I9A.PredictedPass.gff3\", \"9/I9A.gapfilled.final.fa\"),\n",
    "    (\"17/I17V.PredictedPass.gff3\", \"17/I17V.gapfilled.final.fa\"),\n",
    "    (\"19/I19A.PredictedPass.gff3\", \"19/I19A.gapfilled.final.fa\"),\n",
    "    (\"31/I31A.PredictedPass.gff3\", \"31/I31A.gapfilled.final.fa\"),\n",
    "    (\"37/I37A.PredictedPass.gff3\", \"37/I37A.gapfilled.final.fa\"),\n",
    "    (\"61/I61A.PredictedPass.gff3\", \"61/I61A.gapfilled.final.fa\"),\n",
    "    (\"65/I65V.PredictedPass.gff3\", \"65/I65V.gapfilled.final.fa\"),\n",
    "    (\"USR5/I5V.PredictedPass.gff3\", \"USR5/I5V.gapfilled.final.fa\")\n",
    "    ]\n",
    "\n",
    "for in_gff, in_fasta in in_files:\n",
    "    \"\"\" Prepare filenames for IO. \"\"\"\n",
    "    split_gff_path = os.path.splitext(in_gff)[0]\n",
    "    out_cds_file = split_gff_path + cds_suffix\n",
    "    out_orf_file = split_gff_path + orf_suffix\n",
    "    out_dubious_cds_file = split_gff_path + dubious_suffix\n",
    "    out_cleaned_orf_file = split_gff_path + cleaned_suffix + orf_suffix\n",
    "    out_cleaned_cds_file = split_gff_path + cleaned_suffix + cds_suffix\n",
    "    cleaned_gff_file = split_gff_path + cleaned_suffix + '.gff3'\n",
    "\n",
    "    \"\"\" Process the GFF to recode multiple CDS's as exons. \"\"\"\n",
    "    gff_correct(in_gff, cleaned_gff_file)\n",
    "\n",
    "    \"\"\" Extract sequences from gff CDS and exon features. \"\"\"\n",
    "    sequences, dubious_sequences = gff_to_genes(cleaned_gff_file, in_fasta)\n",
    "\n",
    "    \"\"\" Write sequences and dubious sequences to fasta files. \"\"\"\n",
    "    with open(out_cleaned_cds_file, 'w') as handle:\n",
    "        SeqIO.write(sequences, handle, 'fasta')\n",
    "    with open(out_dubious_cds_file, 'w') as handle:\n",
    "        SeqIO.write(dubious_sequences, handle, 'fasta')\n",
    "    translated_sequences = list()\n",
    "    for sequence in sequences:\n",
    "        sequence.seq = sequence.seq.translate()\n",
    "        translated_sequences.append(sequence)\n",
    "    with open(out_cleaned_orf_file, 'w') as handle:\n",
    "        SeqIO.write(translated_sequences, handle, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The predicted genes have now been extracted and the nucleotide and amino acid sequences saved as fasta files.\n",
    "Genes that the program could not extract properly or are not annotated properly (those with unexpected stop codons) are saved to another file with `dubious_suffix` for later follow up.\n",
    "These dubious sequences are still in the other output files."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
